### PR TITLE
Retract of weak equivalences is a weak equivalence

### PR DESCRIPTION
**Pull request by sglasman**

### CODE REVIEW

File: src/category_theory/model/wfs.lean [hunk 1]
```diff
@@ -47,6 +49,18 @@ begin
   refine ⟨r.ib ≫ l, _, _⟩,
   { rw [←category.assoc, ←r.hi, category.assoc, hl₁, ←category.assoc, r.ha, category.id_comp] },
   { rw [category.assoc, hl₂, ←category.assoc, r.hb, category.id_comp] }
+end    
+
+lemma is_wfs.retract_right {L R :morphism_class M } (w : is_wfs L R)
```
Comment (sglasman) [inline src/category_theory/model/wfs.lean:54]: I ended up copying the proof of the previous lemma for this. I first tried to prove it by applying the previous lemma to the opposite category, but I quickly got bogged down.

File: src/category_theory/model/category.lean [hunk 2]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
 -- TODO: Show that it follows that W is closed under retracts. See
 -- https://ncatlab.org/nlab/show/model+category#ClosureOfMorphisms
 
+lemma is_model_category.weq_of_weq_retract_fib { W C F : morphism_class M } ( h : is_model_category W C F )
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:22]: It would be more natural to state this as a "theorem about model categories", rather than a "theorem about classes of maps that form a model category structure", and maybe in a separate file like `category_theory.model.lemmas`. Something like:
```lean
lemma weq_of_weq_retract_fib {M : Type u} [model_category.{v} M] {a b a' b'} ...
```
Currently there aren't any examples of this here because I didn't actually prove any theorems about model categories! But there are a bunch in the lean-homotopy-theory project, e.g. in https://github.com/rwbarton/lean-homotopy-theory/tree/lean-3.4.2/src/homotopy_theory/formal/cofibrations.

There might be some rough edges in the way the `model_category` class is set up given that it's not currently used as a hypothesis anywhere.

File: src/category_theory/model/category.lean [hunk 3]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+ {a b a' b'} {f : a ⟶ b} {f' : a' ⟶ b'} (r : retract f f') (hf : W f) (hf': F f') : W f' := begin
+   rcases h.acf.fact f with ⟨ c, α, β, WCα, Fβ, f_fact ⟩,
+   rw h.acf.llp at WCα,
+   choose l hl using WCα hf' r.ra (β ≫ r.rb) (by { rw [← category.assoc, f_fact], exact r.hr }),
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:26]: `choose` is kind of overkill here. The usual purpose of `choose` is to take a hypothesis of a form like `∀ (x₁ x₂ : X), ∃ (y : Y), P x₁ x₂ y` and produce a choice function `f : X → X → Y`, which needs the axiom of choice.
Here the hypothesis `WCα hf' ...` already has the form `∃ l, P l` with zero `∀`s before it, and the overall goal `W f'` is a proposition. So, you could just extract a witness using `rcases`. In modern mathlib, it's nicer to use `obtain` because it puts the new variables at the start, like `have` or `let`. `choose` also has this advantage, so I suppose it's fine to use it here too.

File: src/category_theory/model/category.lean [hunk 4]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
 -- TODO: Show that it follows that W is closed under retracts. See
 -- https://ncatlab.org/nlab/show/model+category#ClosureOfMorphisms
 
+lemma is_model_category.weq_of_weq_retract_fib { W C F : morphism_class M } ( h : is_model_category W C F )
+ {a b a' b'} {f : a ⟶ b} {f' : a' ⟶ b'} (r : retract f f') (hf : W f) (hf': F f') : W f' := begin
+   rcases h.acf.fact f with ⟨ c, α, β, WCα, Fβ, f_fact ⟩,
+   rw h.acf.llp at WCα,
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:25]: Since you undo this `rw` later, maybe it's more natural to just use `h.acf.lp WCα hf' ...` on the next line instead.
(General principle: try to use lemmas (like `is_wfs.lp`) rather than unfolding definitions. In this library, though, the supporting lemmas that ought to be there might not exist because I only proved what I needed to construct model structures.)

File: src/category_theory/model/category.lean [hunk 5]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+end
+
+#check @Is_pushout.uniqueness
+
+lemma is_model_category.weq_of_weq_retract { W C F : morphism_class M } [ @has_pushouts M 𝓜] ( h : is_model_category W C F )
+ {a b a' b'} {f : a ⟶ b} {f' : a' ⟶ b'} (r : retract f f') (hf : W f) : W f' := begin
+   rcases h.acf.fact f' with ⟨ c, α, β, WCα, Fβ, f'_fact ⟩,
+   cases has_pushouts.pushout α r.ia with z γ δ po,
+   have WCδ : (C ∩ W) δ := by { rw h.acf.llp, rw h.acf.llp at WCα, exact llp_pushout F po WCα },
+   have uq := @Is_pushout.uniqueness M 𝓜 a' c a z α r.ia γ δ po,
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:49]: This doesn't actually appear to be necessary: it worked fine if I replaced the single use of `uq` below by `po.uniqueness`.

File: src/category_theory/model/category.lean [hunk 6]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+   have rεβ : retract (po.induced (β ≫ r.ib) f (by { rw [← category.assoc, f'_fact], exact r.hi.symm})) β := {
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:50]: I would probably name this map `po.induced ...` with a `let` for clarity, even though you never actually need to refer to it later.

File: src/category_theory/model/category.lean [hunk 7]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+     ra := po.induced (𝟙 c) (r.ra ≫ α) (by { rw [category.comp_id, ← category.assoc, r.ha, category.id_comp] }),
+     ib := r.ib,
+     rb := r.rb,
+     ha := by simp,
+     hb := r.hb,
+     hi := by simp,
+     hr := by { 
+      refine uq _ _, 
+      rw [← category.assoc,
+          Is_pushout.induced_commutes₀ po _ _ _,
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:61]: Since `po : Is_pushout ...`, you can abbreviate this to `po.induced_commutes₀`.

File: src/category_theory/model/category.lean [hunk 8]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+     ia := γ,
+     ra := po.induced (𝟙 c) (r.ra ≫ α) (by { rw [category.comp_id, ← category.assoc, r.ha, category.id_comp] }),
+     ib := r.ib,
+     rb := r.rb,
+     ha := by simp,
+     hb := r.hb,
+     hi := by simp,
+     hr := by { 
+      refine uq _ _, 
+      rw [← category.assoc,
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:60]: It's a good habit to use `{ }` whenever there are multiple goals, unless you want to treat multiple goals at once with something like `all_goals`. Even when you're going to kill each goal with a single tactic anyways, it makes the structure of the proof more apparent to the reader.

File: src/category_theory/model/category.lean [hunk 9]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+   have uq := @Is_pushout.uniqueness M 𝓜 a' c a z α r.ia γ δ po,
+   have rεβ : retract (po.induced (β ≫ r.ib) f (by { rw [← category.assoc, f'_fact], exact r.hi.symm})) β := {
+     ia := γ,
+     ra := po.induced (𝟙 c) (r.ra ≫ α) (by { rw [category.comp_id, ← category.assoc, r.ha, category.id_comp] }),
+     ib := r.ib,
+     rb := r.rb,
+     ha := by simp,
+     hb := r.hb,
+     hi := by simp,
+     hr := by {
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:58]: As a style thing, we normally only use `by` for things that fit on one line--for longer blocks `begin`/`end` is clearer (it's easier to distinguish from other meanings of `{}`).

File: src/category_theory/model/category.lean [hunk 10]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+     hb := r.hb,
+     hi := by { rw [category.assoc, f_fact], exact r.hi },
+     hr := hl.2,
+   },
+   rw ← h.acf.llp at WCα,
+   rw ← f_fact at hf,
+   exact (is_wfs.retract_right h.caf rβf' ⟨ Fβ, h.weq.weq_cancel_left WCα.2 hf ⟩).2
+end
+
+#check @Is_pushout.uniqueness
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:42]: This seems to be left over.

File: src/category_theory/model/category.lean [hunk 11]
```diff
@@ -19,14 +19,75 @@ structure is_model_category (W C F : morphism_class M) : Prop :=
...
+
+lemma is_model_category.weq_of_weq_retract { W C F : morphism_class M } [ @has_pushouts M 𝓜] ( h : is_model_category W C F )
+ {a b a' b'} {f : a ⟶ b} {f' : a' ⟶ b'} (r : retract f f') (hf : W f) : W f' := begin
+   rcases h.acf.fact f' with ⟨ c, α, β, WCα, Fβ, f'_fact ⟩,
+   cases has_pushouts.pushout α r.ia with z γ δ po,
+   have WCδ : (C ∩ W) δ := by { rw h.acf.llp, rw h.acf.llp at WCα, exact llp_pushout F po WCα },
```
Comment (rwbarton) [inline src/category_theory/model/category.lean:48]: You can abbreviate the start of this proof to `rw h.acf.llp at ⊢ WCo` (`⊢` is `\|-`).

File: src/category_theory/model/wfs.lean (hunk 1 above)
Comment (rwbarton) [reply src/category_theory/model/wfs.lean:54]: Yes. It was probably made more difficult by the way that the opposite category was implemented at the time. We later found that it was very easy to get confused about whether we were considering objects and morphisms in a category or in its opposite. Nowadays we have a design in which you need to apply specific functions to convert between a category and its opposite, which helps keep things straight.

Ideally we will some day have a method to automatically dualize theorems, but this doesn't exist yet.

**Review by rwbarton — COMMENTED** [review]
Thanks Saul, this looks great! The proofs are nice and direct. I left some style-related comments. Also, check out https://leanprover-community.github.io/contribute/style.html#structuring-definitions-and-theorems for the standard formatting around things like parentheses/braces (no spaces except inside `{ }` for records and tactic blocks), starting proofs on the next line, indenting long lemma statements and so on. Generally I try to stick to mathlib formatting style even in non-mathlib projects.

File: src/category_theory/model/category.lean (hunk 2 above)
Comment (sglasman) [reply src/category_theory/model/category.lean:22]: So I have a question about this. Trying to rephrase this to use `model_category.{v} M` rather than `is_model_category W C F` (and omitting \MCM from the context) I end up with 
```
lemma weq_of_weq_retract_fib [ h : model_category.{v} M ]
 {a b a' b': M} {f : a ⟶ b} {f' : a' ⟶ b'} (r : retract f f') (hf : h.W f) (hf': h.F f') : h.W f'
```
but `h.W` and `h.F` produce the following error:

"invalid field notation, function 'category_theory.model_category.W' does not have explicit argument with type (category_theory.model_category ...)".

I don't understand this - if I `#check h.W`, it's a morphism class on M like I expect.

File: src/category_theory/model/category.lean (hunk 3 above)
Comment (sglasman) [reply src/category_theory/model/category.lean:26]: I could have sworn I tried `rcases` first, but it does work. I've never encountered `obtain`, I'll look it up

File: src/category_theory/model/category.lean (hunk 6 above)
Comment (sglasman) [reply src/category_theory/model/category.lean:50]: Ah great - I was trying to do this with `have` rather than `let`, but `have` appears to only remember the type, not the definition
